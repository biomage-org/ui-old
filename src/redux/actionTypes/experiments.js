const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Experiments loaded
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Example experiments loaded
 */
const EXPERIMENTS_EXAMPLES_LOADED = `${EXPERIMENTS}/examplesLoaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

const EXPERIMENTS_SAVED = `${EXPERIMENTS}/saved`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

/**
 * Set active project.
 */
const EXPERIMENTS_SET_ACTIVE = `${EXPERIMENTS}/setActive`;

/**
 * Add project metadata.
 */
const EXPERIMENTS_METADATA_CREATE = `${EXPERIMENTS}/metadataCreate`;

/**
 * Update project metadata.
 */
const EXPERIMENTS_METADATA_RENAME = `${EXPERIMENTS}/metadataRename`;

/**
 * Delete project metadta.
 */
const EXPERIMENTS_METADATA_DELETE = `${EXPERIMENTS}/metadataDelete`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_EXAMPLES_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
  EXPERIMENTS_SWITCH,
  EXPERIMENTS_SET_ACTIVE,
  EXPERIMENTS_METADATA_CREATE,
  EXPERIMENTS_METADATA_RENAME,
  EXPERIMENTS_METADATA_DELETE,
};
