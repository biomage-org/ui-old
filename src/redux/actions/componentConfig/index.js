import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import getDotPlot from 'redux/actions/componentConfig/getDotPlot';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  getDotPlot,
  updatePlotData,
  resetPlotConfig,
};
