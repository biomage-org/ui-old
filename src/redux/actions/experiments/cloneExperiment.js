/* eslint-disable no-param-reassign */
import fetchAPI from 'utils/http/fetchAPI';
import moment from 'moment';

import {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_SAVING,
} from 'redux/actionTypes/experiments';

import endUserMessages from 'utils/endUserMessages';

import handleError from 'utils/http/handleError';

const cloneExperiment = (
  originalId, name = null,
) => async (dispatch) => {
  dispatch({
    type: EXPERIMENTS_SAVING,
  });

  try {
    const createdAt = moment().toISOString();

    const newExperimentId = await fetchAPI(`/v2/experiments/${originalId}/clone`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name }),
    });

    dispatch({
      type: EXPERIMENTS_CREATED,
      payload: {
        experiment: { id: newExperimentId, name, createdAt },
      },
    });

    return newExperimentId;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_SAVING);

    dispatch({
      type: EXPERIMENTS_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }
};

export default cloneExperiment;
