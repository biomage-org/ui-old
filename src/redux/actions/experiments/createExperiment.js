/* eslint-disable no-param-reassign */
import fetchAPI from 'utils/http/fetchAPI';
import moment from 'moment';
import hash from 'object-hash';
import {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_SAVING,
} from 'redux/actionTypes/experiments';

import endUserMessages from 'utils/endUserMessages';

import handleError from 'utils/http/handleError';

const createExperiment = (
  name, description,
) => async (dispatch) => {
  const createdAt = moment().toISOString();
  const experimentId = hash.MD5(createdAt);

  const newExperimentProperties = {
    id: experimentId,
    name,
    description,
  };

  dispatch({
    type: EXPERIMENTS_SAVING,
  });

  try {
    await fetchAPI(
      `/v2/experiments/${experimentId}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(newExperimentProperties),
      },
    );

    dispatch({
      type: EXPERIMENTS_CREATED,
      payload: {
        // TODO We don't really need to send this createdAt to redux, the real createdAt
        // is being generated in the api
        // We should make the POST to create the experiment return the new experiment and
        // Take the createdAt from there
        experiment: { createdAt, ...newExperimentProperties },
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_SAVING);

    dispatch({
      type: EXPERIMENTS_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }

  return Promise.resolve(experimentId);
};

export default createExperiment;
