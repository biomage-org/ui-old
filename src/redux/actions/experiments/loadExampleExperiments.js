import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import endUserMessages from 'utils/endUserMessages';

import { EXPERIMENTS_ERROR, EXPERIMENTS_EXAMPLES_LOADED, EXPERIMENTS_LOADING } from 'redux/actionTypes/experiments';

const loadExampleExperiments = () => async (dispatch) => {
  dispatch({
    type: EXPERIMENTS_LOADING,
  });

  try {
    const experiments = await fetchAPI('/v2/experiments/examples');

    dispatch({
      type: EXPERIMENTS_EXAMPLES_LOADED,
      payload: {
        experiments,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_LOADING_PROJECT);

    dispatch({
      type: EXPERIMENTS_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }
};

export default loadExampleExperiments;
