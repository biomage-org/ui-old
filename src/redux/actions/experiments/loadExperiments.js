import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import endUserMessages from 'utils/endUserMessages';

import { EXPERIMENTS_ERROR, EXPERIMENTS_LOADED, EXPERIMENTS_LOADING } from 'redux/actionTypes/experiments';

const loadExperiments = () => async (dispatch) => {
  dispatch({
    type: EXPERIMENTS_LOADING,
  });

  try {
    const experiments = await fetchAPI('/v2/experiments');
    experiments.forEach((experiment) => {
      // eslint-disable-next-line no-param-reassign
      experiment.sampleIds = experiment.samplesOrder;
      // eslint-disable-next-line no-param-reassign
      delete experiment.samplesOrder;
    });

    dispatch({
      type: EXPERIMENTS_LOADED,
      payload: {
        experiments,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_LOADING_PROJECT);

    dispatch({
      type: EXPERIMENTS_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }
};
export default loadExperiments;
