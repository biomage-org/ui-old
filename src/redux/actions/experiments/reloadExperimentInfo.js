import { EXPERIMENT_SETTINGS_INFO_UPDATE } from 'redux/actionTypes/experimentSettings';
import fetchAPI from 'utils/http/fetchAPI';

const toApiV1 = (experimentV2) => {
  const {
    id, name, samplesOrder, pipelines, ...restOfExperiment
  } = experimentV2;

  const pipelinesV1 = {
    gem2s: pipelines.gem2s,
    pipeline: pipelines.qc,
  };

  const experimentV1 = {
    ...restOfExperiment,
    experimentId: id,
    experimentName: name,
    projectId: id,
    sampleIds: samplesOrder,
    meta: pipelinesV1,
    // These are always created with the same value right now
    // when the UI is updated:
    //  - Organism is not going to be used anymore.
    //  - Type will be defined in the samples, not in the experiment.
    organism: null,
    type: '10x',
  };

  return experimentV1;
};

const reloadExperimentInfo = () => async (dispatch, getState) => {
  const { experimentId } = getState().experimentSettings.info;

  const experimentInfoV2 = await fetchAPI(`/v2/experiments/${experimentId}`);

  dispatch({
    type: EXPERIMENT_SETTINGS_INFO_UPDATE,
    payload: toApiV1(experimentInfoV2),
  });
};

export default reloadExperimentInfo;
