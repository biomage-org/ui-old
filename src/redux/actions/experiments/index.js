import createExperiment from './createExperiment';
import deleteExperiment from './deleteExperiment';
import updateExperiment from './updateExperiment';
import loadExperiments from './loadExperiments';
import setActiveExperiment from './setActiveExperiment';
import switchExperiment from './switchExperiment';
import reorderSamples from './reorderSamples';
import createMetadataTrack from './createMetadataTrack';
import renameMetadataTrack from './renameMetadataTrack';
import deleteMetadataTrack from './deleteMetadataTrack';
import updateValueInMetadataTrack from './updateValueInMetadataTrack';
import cloneExperiment from './cloneExperiment';
import reloadExperimentInfo from './reloadExperimentInfo';

export {
  createExperiment,
  deleteExperiment,
  updateExperiment,
  loadExperiments,
  setActiveExperiment,
  switchExperiment,
  reorderSamples,
  createMetadataTrack,
  renameMetadataTrack,
  deleteMetadataTrack,
  updateValueInMetadataTrack,
  cloneExperiment,
  reloadExperimentInfo,
};
