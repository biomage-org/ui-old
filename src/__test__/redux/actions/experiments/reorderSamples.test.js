import _ from 'lodash';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import {
  EXPERIMENTS_ERROR,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_UPDATED,
} from 'redux/actionTypes/experiments';
import { reorderSamples } from 'redux/actions/experiments';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

const mockStore = configureStore([thunk]);

enableFetchMocks();

describe('updateExperiment', () => {
  const experimentId = 'experiment-1';
  const oldIndex = 0;
  const newIndex = 1;
  const newSampleOrder = ['sample2', 'sample1'];

  const mockExperiment = {
    ...experimentTemplate,
    name: 'experiment-1',
    id: experimentId,
  };

  const mockState = {
    experiments: {
      ...initialExperimentState,
      [experimentId]: mockExperiment,
    },
  };

  beforeEach(() => {
    fetchMock.resetMocks();
    fetchMock.doMock();
  });

  it('Updates properties correctly', async () => {
    const response = new Response(JSON.stringify(newSampleOrder));
    fetchMock.mockResolvedValueOnce(response);

    const store = mockStore(mockState);
    await store.dispatch(reorderSamples(experimentId, oldIndex, newIndex));

    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([EXPERIMENTS_SAVING, EXPERIMENTS_UPDATED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();

    expect(fetchMock).toHaveBeenCalledWith(
      `http://localhost:3000/v2/experiments/${experimentId}/samples/position`,
      expect.objectContaining({
        method: 'PUT',
        body: JSON.stringify({
          oldPosition: oldIndex,
          newPosition: newIndex,
        }),
      }),
    );
  });

  it('Handles error and then rethrows (so the UI doesnt reorder)', async () => {
    const error = new Error('Some api error');

    fetchMock.mockRejectOnce(() => Promise.reject(error));

    const store = mockStore(mockState);
    await expect(
      store.dispatch(reorderSamples(experimentId, oldIndex, newIndex)),
    ).rejects.toThrow(new Error(`Error: ${error.message}`));

    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([EXPERIMENTS_SAVING, EXPERIMENTS_ERROR]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();

    expect(fetchMock).toHaveBeenCalledWith(
      `http://localhost:3000/v2/experiments/${experimentId}/samples/position`,
      expect.objectContaining({
        method: 'PUT',
        body: JSON.stringify({
          oldPosition: oldIndex,
          newPosition: newIndex,
        }),
      }),
    );
  });
});
