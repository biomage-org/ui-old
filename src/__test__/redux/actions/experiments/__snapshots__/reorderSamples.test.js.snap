// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateExperiment Handles error and then rethrows (so the UI doesnt reorder) 1`] = `
[
  undefined,
  {
    "error": "Error: Some api error",
  },
]
`;

exports[`updateExperiment Updates properties correctly 1`] = `
[
  undefined,
  {
    "experiment": {
      "sampleIds": [
        "sample2",
        "sample1",
      ],
    },
    "experimentId": "experiment-1",
  },
]
`;
